# **Pandas**
Pandas bir Python kütüphanesidir.

Pandas veri analizi için kullanılır.

**Pandas'ı Neden Kullanmalıyız?**

Pandas, büyük verileri analiz etmemize ve istatistiksel teorilere dayanarak sonuçlar çıkarmamıza olanak tanır.

Pandas, karmaşık veri kümelerini temizleyebilir, okunabilir ve anlamlı hale getirebilir.

Veri biliminde ilgili veriler çok önemlidir.

Veri Bilimi: Verilerin nasıl depolandığını, kullanıldığını ve analiz edildiğini, böylece bunlardan bilgi elde edildiğini inceleyen bilgisayar biliminin bir dalıdır.

**Pandalar Neler Yapabilir?**

Pandas size veriler hakkında cevaplar verir. Örneğin:

İki veya daha fazla sütun arasında bir ilişki var mı?
Ortalama değer nedir?
Maksimum değer?
Minimum değer?
Pandas ayrıca alakasız veya yanlış değerler (örneğin boş veya NULL değerler) içeren satırları da silebilir. Buna veri temizleme denir.




In [1]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


Pandas genellikle takma ad altında içe aktarılır pd .

Python'da takma adlar, aynı şeye atıfta bulunmak için kullanılan alternatif isimlerdir.

asİçe aktarma sırasında anahtar kelimeyle bir takma ad oluşturun :

In [2]:
#Sürüm dizesi öznitelik altında saklanır __version__ .

import pandas as pd

print(pd.__version__)

2.2.2


# **Dizi nedir?**

Pandas Serisi, bir tablodaki sütun gibidir.

Herhangi bir türdeki veriyi tutabilen tek boyutlu bir dizidir.

In [ ]:
#Bir listeden basit bir Pandas Serisi oluşturun:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

# **Etiketler**
Başka bir şey belirtilmediği takdirde, değerler indeks numaralarıyla etiketlenir. Birinci değerin indeksi 0, ikinci değerin indeksi 1 vb. şeklindedir.

Bu etiket, belirtilen bir değere erişmek için kullanılabilir.

Etiketler Oluştur
Bu argümanla indexkendi etiketlerinizi adlandırabilirsiniz.

In [3]:
#Kendi etiketlerinizi oluşturun:

import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


Seri Olarak Anahtar/Değer Nesneleri

Seriler oluştururken sözlük gibi anahtar/değer nesnesi de kullanabilirsiniz.

In [4]:
#Sözlükten basit bir Pandas Serisi oluşturun:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)

day1    420
day2    380
day3    390
dtype: int64


Not: Sözlüğün anahtarları etiketlere dönüşür.

Sözlükteki öğelerin yalnızca bir kısmını seçmek için, index bağımsız değişkeni kullanın ve Seriye dahil etmek istediğiniz öğeleri belirtin.

In [5]:
#Yalnızca "1. gün" ve "2. gün" verilerini kullanarak bir Seri oluşturun:

import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


Veri Çerçeveleri

Pandas'taki veri kümeleri genellikle DataFrame olarak adlandırılan çok boyutlu tablolardır.

Seriler bir sütun gibidir, DataFrame ise tüm tablodur.

In [6]:
#İki Seriden bir Veri Çerçevesi oluşturun:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)


   calories  duration
0       420        50
1       380        40
2       390        45


DataFrame nedir?

Pandas DataFrame, iki boyutlu bir dizi veya satır ve sütunlardan oluşan bir tablo gibi, iki boyutlu bir veri yapısıdır.

In [7]:
#Basit bir Pandas DataFrame oluşturun:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


In [8]:
#0. satırı döndür:

print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [9]:
#0. ve 1. satırları döndürün:

print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


In [10]:
#Her satıra bir ad vermek için bir ad listesi ekleyin:

import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [11]:
#"day2"yi döndür:

print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


# Dosyaları Veri Çerçevesine Yükle

Veri kümeleriniz bir dosyada saklanıyorsa, Pandas bunları bir DataFrame'e yükleyebilir.

In [12]:
#Virgülle ayrılmış bir dosyayı (CSV dosyası) bir DataFrame'e yükleyin:

import pandas as pd

df = pd.read_csv('data.csv')

print(df)

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

# CSV Dosyalarını Okuyun

Büyük veri kümelerini saklamanın basit bir yolu CSV dosyalarını (virgülle ayrılmış dosyalar) kullanmaktır.

CSV dosyaları düz metin içerir ve Pandas dahil herkes tarafından okunabilen, iyi bilinen bir formattır.

Örneklerimizde 'data.csv' adlı bir CSV dosyası kullanacağız.

In [13]:
#CSV dosyasını bir DataFrame'e yükleyin:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

In [22]:
#Eğer çok sayıda satır içeren büyük bir DataFrame'iniz varsa, Pandas yalnızca ilk 5 satırı ve son 5 satırı döndürür:

#DataFrame'i to_string() metot olmadan yazdırın:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

In [18]:
#Döndürülen maksimum satır sayısını kontrol edin:

import pandas as pd

print(pd.options.display.max_rows)

60


In [20]:
#DataFrame'in tamamını görüntülemek için maksimum satır sayısını artırın:

import pandas as pd

pd.options.display.max_rows = 9999

df = pd.read_csv('data.csv')

print(df)

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

# **JSON'u okuyun**

Büyük veri kümeleri genellikle JSON formatında saklanır veya çıkarılır.

JSON düz metindir, ancak nesne biçimine sahiptir ve Pandas dahil olmak üzere programlama dünyasında iyi bilinir.

Örneklerimizde 'data.json' adlı bir JSON dosyası kullanacağız.

In [21]:
#JSON dosyasını bir DataFrame'e yükleyin:

import pandas as pd

df = pd.read_json('data.json')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

In [23]:
#Python sözlüğünü bir DataFrame'e yükleme:

import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


# **Verileri Görüntüleme**

DataFrame'in içeriğine hızlıca genel bir bakış elde etmek için en çok kullanılan yöntemlerden biri de bu head()yöntemdir.

Bu head()yöntem, en üstten başlayarak başlıkları ve belirtilen sayıda satırı döndürür.

In [24]:
#DataFrame'in ilk 10 satırını yazdırarak hızlı bir genel bakış elde edin:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.head(10))

          Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105
5          VW         Up!    1000     929  105
6       Skoda       Fabia    1400    1109   90
7    Mercedes     A-Class    1500    1365   92
8        Ford      Fiesta    1500    1112   98
9        Audi          A1    1600    1150   99


In [25]:
#Veri çerçevesinin ilk 5 satırını yazdırın:

import pandas as pd

df = pd.read_csv('data.csv')

print(df.head())

          Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105


In [26]:
#Veri çerçevesinin son 5 satırını yazdırın:

print(df.tail())

         Car   Model  Volume  Weight  CO2
31     Volvo    XC70    2000    1746  117
32      Ford   B-Max    1600    1235  104
33       BMW     216    1600    1390  108
34      Opel  Zafira    1600    1405  109
35  Mercedes     SLK    2500    1395  120


In [27]:
#Verilerle ilgili bilgileri yazdırın:

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Car     36 non-null     object
 1   Model   36 non-null     object
 2   Volume  36 non-null     int64 
 3   Weight  36 non-null     int64 
 4   CO2     36 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 1.5+ KB
None


# **Veri Temizleme**
Veri temizleme, veri setinizdeki hatalı verileri düzeltmek anlamına gelir.

Hatalı veriler şunlar olabilir:

Boş hücreler
Veriler yanlış formatta.
Yanlış veri
Kopyalar
Bu eğitimde bunların hepsiyle nasıl başa çıkacağınızı öğreneceksiniz.

Veri Setimiz
Sonraki bölümlerde bu veri setini kullanacağız:


      Süre Tarih Nabız Maksimum Nabız Kalori
  0 60 '2020/12/01' 110 130 409.1
  1 60 '2020/12/02' 117 145 479.0
  2 60 '2020/12/03' 103 135 340.0
  3 45 '2020/12/04' 109 175 282.4
  4 45 '2020/12/05' 117 148 406.0
  5 60 '2020/12/06' 102 127 300.0
  6 60 '2020/12/07' 110 136 374.0
  7 450 '2020/12/08' 104 134 253.3
  8 30 '2020/12/09' 109 133 195.1
  9 60 '2020/12/10' 98 124 269.0
  10 60 '2020/12/11' 103 147 329.3
  11 60 '2020/12/12' 100 120 250.7
  12 60 '2020/12/12' 100 120 250.7
  13 60 '2020/12/13' 106 128 345.3
  14 60 '2020/12/14' 104 132 379.3
  15 60 '2020/12/15' 98 123 275.0
  16 60 '2020/12/16' 98 120 215.2
  17 60 '2020/12/17' 100 120 300.0
  18 45 '2020/12/18' 90 112 NaN
  19 60 '2020/12/19' 103 123 323.0
  20 45 '2020/12/20' 97 125 243.0
  21 60 '2020/12/21' 108 131 364.2
  22 45 NaN 100 119 282.0
  23 60 '2020/12/23' 130 101 300.0
  24 45 '2020/12/24' 105 132 246.0
  25 60 '2020/12/25' 102 126 334.5
  26 60 2020/12/26 100 120 250.0
  27 60 '2020/12/27' 92 118 241.0
  28 60 '2020/12/28' 103 132 NaN
  29 60 '2020/12/29' 100 132 280.0
  30 60 '2020/12/30' 102 129 380.3
  31 60 '2020/12/31' 92 115 243.0

Veri kümesinde bazı boş hücreler bulunmaktadır ("Tarih" 22. satırda ve "Kalori" 18. ve 28. satırlarda).

Veri kümesinde yanlış format bulunmaktadır ("Tarih" 26. satırda).

Veri kümesi hatalı veri içeriyor (7. satırda "Süre").

Veri kümesinde tekrarlanan kayıtlar bulunmaktadır (11. ve 12. satırlar).

# Boş Hücreler

Boş hücreler, verileri analiz ederken yanlış sonuçlara yol açabilir.

Satırları Kaldır
Boş hücrelerle başa çıkmanın bir yolu, boş hücre içeren satırları silmektir.

Bu genellikle sorun teşkil etmez, çünkü veri kümeleri çok büyük olabilir ve birkaç satırın kaldırılması sonuç üzerinde büyük bir etki yaratmaz.

In [28]:
#Boş hücre içermeyen yeni bir Veri Çerçevesi döndürün:

import pandas as pd

df = pd.read_csv('data.csv')

new_df = df.dropna()

print(new_df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

Not: Varsayılan olarak, bu yöntem yenidropna() bir DataFrame döndürür ve orijinalini değiştirmez.

Orijinal DataFrame'i değiştirmek istiyorsanız şu inplace = Trueargümanı kullanın:

In [29]:
#Boş (NULL) değer içeren tüm satırları silin:

import pandas as pd

df = pd.read_csv('data.csv')

df.dropna(inplace = True)

print(df.to_string())

           Car       Model  Volume  Weight  CO2
0       Toyoty        Aygo    1000     790   99
1   Mitsubishi  Space Star    1200    1160   95
2        Skoda      Citigo    1000     929   95
3         Fiat         500     900     865   90
4         Mini      Cooper    1500    1140  105
5           VW         Up!    1000     929  105
6        Skoda       Fabia    1400    1109   90
7     Mercedes     A-Class    1500    1365   92
8         Ford      Fiesta    1500    1112   98
9         Audi          A1    1600    1150   99
10     Hyundai         I20    1100     980   99
11      Suzuki       Swift    1300     990  101
12        Ford      Fiesta    1000    1112   99
13       Honda       Civic    1600    1252   94
14      Hundai         I30    1600    1326   97
15        Opel       Astra    1600    1330   97
16         BMW           1    1600    1365   99
17       Mazda           3    2200    1280  104
18       Skoda       Rapid    1600    1119  104
19        Ford       Focus    2000    13

Not: Bu işlem dropna(inplace = True)yeni bir DataFrame döndürmeyecek, ancak orijinal DataFrame'den NULL değerleri içeren tüm satırları kaldıracaktır.

# Boş Değerleri Değiştir
Boş hücrelerle başa çıkmanın bir başka yolu da, bunun yerine yeni bir değer eklemektir .

Bu sayede, sadece birkaç boş hücre yüzünden tüm satırları silmek zorunda kalmazsınız.

Bu fillna()yöntem, boş hücreleri bir değerle değiştirmemizi sağlar:

In [30]:
#Boş değerleri 130 sayısıyla değiştirin:

import pandas as pd

df = pd.read_csv('data.csv')

df.fillna(130, inplace = True)

# Yalnızca Belirtilen Sütunlar İçin Değiştir
Yukarıdaki örnek, tüm Veri Çerçevesindeki boş hücrelerin tamamını değiştirir.

Yalnızca bir sütundaki boş değerleri değiştirmek için, DataFrame'in sütun adını belirtin:

In [31]:
#"Kalori" sütunundaki boş (NULL) değerleri 130 sayısıyla değiştirin:

import pandas as pd

df = pd.read_csv('data.csv')

df.fillna({"Calories": 130}, inplace=True)

# Ortalama, Medyan veya Mod Kullanarak Değiştirin
Boş hücreleri doldurmanın yaygın bir yolu, sütunun ortalama, medyan veya mod değerini hesaplamaktır.

Pandas, belirtilen bir sütun için ilgili değerleri hesaplamak üzere mean() median()`and` yöntemlerini kullanır :mode()

In [36]:
#Ortalamayı hesaplayın ve boş kalan değerleri bu ortalama ile değiştirin:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Weight"].mean()

print(df.fillna({"Weight": x}, inplace=True))

None


Medyan = tüm değerler artan sırada sıralandıktan sonra ortadaki değer.

In [35]:
#Modu hesaplayın ve boş olan değerleri bu mod ile değiştirin:

import pandas as pd

df = pd.read_csv('data.csv')

x = df["Weight"].mode()[0]

print(df.fillna({"Weight": x}, inplace=True))



None


# **Veriler yanlış formatta**
Veri formatı yanlış olan hücreler, verilerin analizini zorlaştırabilir, hatta imkansız hale getirebilir.

Bunu düzeltmek için iki seçeneğiniz var: satırları silmek veya sütunlardaki tüm hücreleri aynı biçime dönüştürmek.

Doğru Biçime Dönüştür
Veri çerçevemizde, yanlış biçimde yazılmış iki hücre var. 22. ve 26. satırlara bakın, 'Tarih' sütunu bir tarihi temsil eden bir metin olmalıdır:


      Süre Tarih Nabız Maksimum Nabız Kalori
  0 60 '2020/12/01' 110 130 409.1
  1 60 '2020/12/02' 117 145 479.0
  2 60 '2020/12/03' 103 135 340.0
  3 45 '2020/12/04' 109 175 282.4
  4 45 '2020/12/05' 117 148 406.0
  5 60 '2020/12/06' 102 127 300.0
  6 60 '2020/12/07' 110 136 374.0
  7 450 '2020/12/08' 104 134 253.3
  8 30 '2020/12/09' 109 133 195.1
  9 60 '2020/12/10' 98 124 269.0
  10 60 '2020/12/11' 103 147 329.3
  11 60 '2020/12/12' 100 120 250.7
  12 60 '2020/12/12' 100 120 250.7
  13 60 '2020/12/13' 106 128 345.3
  14 60 '2020/12/14' 104 132 379.3
  15 60 '2020/12/15' 98 123 275.0
  16 60 '2020/12/16' 98 120 215.2
  17 60 '2020/12/17' 100 120 300.0
  18 45 '2020/12/18' 90 112 NaN
  19 60 '2020/12/19' 103 123 323.0
  20 45 '2020/12/20' 97 125 243.0
  21 60 '2020/12/21' 108 131 364,2
  22 45 NaN 100 119 282,0
  23 60 '2020/12/23' 130 101 300.0
  24 45 '2020/12/24' 105 132 246.0
  25 60 '2020/12/25' 102 126 334.5
  26 60 20201226 100 120 250.0
  27 60 '2020/12/27' 92 118 241.0
  28 60 '2020/12/28' 103 132 NaN
  29 60 '2020/12/29' 100 132 280.0
  30 60 '2020/12/30' 102 129 380.3
  31 60 '2020/12/31' 92 115 243.0

'Tarih' sütunundaki tüm hücreleri tarihe dönüştürmeyi deneyelim.

Pandas'ın to_datetime()bunun için bir yöntemi var:

In [ ]:
#Tarihe dönüştür:

#import pandas as pd

#df = pd.read_csv('data.csv')

#df['Date'] = pd.to_datetime(df['Date'], format='mixed')

#print(df.to_string())

Sonuçtan da görebileceğiniz gibi, 26. satırdaki tarih düzeltildi, ancak 22. satırdaki boş tarih NaT (Zaman Değil) değeri, yani boş bir değer aldı. Boş değerlerle başa çıkmanın bir yolu, tüm satırı silmektir.

# **Satırları Kaldırmak**
Yukarıdaki örnekte dönüştürme işleminden elde ettiğimiz sonuç NaT değeridir; bu değer NULL değeri olarak ele alınabilir ve ilgili yöntemi kullanarak satırı kaldırabiliriz dropna().

In [38]:
#"Tarih" sütununda NULL değeri olan satırları kaldırın:

#df.dropna(subset=['Date'], inplace = True)

# **Yanlış Veri**
"Yanlış veri" illa "boş hücreler" veya "yanlış format" anlamına gelmez, sadece yanlış olabilir; örneğin birisi "1.99" yerine "199" kaydetmiş olabilir.

Bazen veri setine bakarak yanlış verileri tespit edebilirsiniz, çünkü verilerin nasıl olması gerektiğine dair bir beklentiniz vardır.

Veri setimize bakarsanız, 7. satırda sürenin 450 olduğunu, ancak diğer tüm satırlarda sürenin 30 ile 60 arasında olduğunu görebilirsiniz.

Yanlış olmak zorunda değil, ancak bunun bir kişinin egzersiz seanslarına ait veri seti olduğunu göz önünde bulundurarak, bu kişinin 450 dakika egzersiz yapmadığı sonucuna varıyoruz.


      Süre Tarih Nabız Maksimum Nabız Kalori
  0 60 '2020/12/01' 110 130 409.1
  1 60 '2020/12/02' 117 145 479.0
  2 60 '2020/12/03' 103 135 340.0
  3 45 '2020/12/04' 109 175 282.4
  4 45 '2020/12/05' 117 148 406.0
  5 60 '2020/12/06' 102 127 300.0
  6 60 '2020/12/07' 110 136 374.0
  7 450 '2020/12/08' 104 134 253.3
  8 30 '2020/12/09' 109 133 195.1
  9 60 '2020/12/10' 98 124 269.0
  10 60 '2020/12/11' 103 147 329.3
  11 60 '2020/12/12' 100 120 250.7
  12 60 '2020/12/12' 100 120 250.7
  13 60 '2020/12/13' 106 128 345.3
  14 60 '2020/12/14' 104 132 379.3
  15 60 '2020/12/15' 98 123 275.0
  16 60 '2020/12/16' 98 120 215.2
  17 60 '2020/12/17' 100 120 300.0
  18 45 '2020/12/18' 90 112 NaN
  19 60 '2020/12/19' 103 123 323.0
  20 45 '2020/12/20' 97 125 243.0
  21 60 '2020/12/21' 108 131 364.2
  22 45 NaN 100 119 282.0
  23 60 '2020/12/23' 130 101 300.0
  24 45 '2020/12/24' 105 132 246.0
  25 60 '2020/12/25' 102 126 334.5
  26 60 20201226 100 120 250.0
  27 60 '2020/12/27' 92 118 241.0
  28 60 '2020/12/28' 103 132 NaN
  29 60 '2020/12/29' 100 132 280.0
  30 60 '2020/12/30' 102 129 380.3
  31 60 '2020/12/31' 92 115 243.0

7. satırdaki "Süre" değeri gibi yanlış değerleri nasıl düzeltebiliriz?

# **Değerlerin Değiştirilmesi**
Yanlış değerleri düzeltmenin bir yolu, onları başka bir şeyle değiştirmektir.

Örneğimizde, büyük olasılıkla bir yazım hatası var ve değer "450" ​​yerine "45" olmalı; 7. satıra da "45" yazabiliriz:

In [ ]:
#7. satırda "Süre" değerini 45 olarak ayarlayın:

#df.loc[7, 'Duration'] = 45

Küçük veri kümelerinde hatalı verileri tek tek değiştirebilirsiniz, ancak büyük veri kümelerinde bu mümkün değildir.

Daha büyük veri kümelerindeki hatalı verileri değiştirmek için bazı kurallar oluşturabilirsiniz; örneğin, geçerli değerler için bazı sınırlar belirleyebilir ve bu sınırların dışında kalan değerleri değiştirebilirsiniz.

In [ ]:
#"Süre" sütunundaki tüm değerleri döngüyle kontrol edin.

#Değer 120'den yüksekse, 120 olarak ayarlayın:

#for x in df.index:
  #if df.loc[x, "Duration"] > 120:
    #df.loc[x, "Duration"] = 120

# **Satırları Kaldırmak**
Yanlış verilerle başa çıkmanın bir başka yolu da yanlış veri içeren satırları silmektir.

Bu sayede onların yerine ne koyacağınızı bulmanıza gerek kalmaz ve analizlerinizi yapmak için onlara ihtiyacınız olma olasılığı da düşüktür.

In [ ]:
#"Süre" değeri 120'den yüksek olan satırları silin:

#for x in df.index:
  #if df.loc[x, "Duration"] > 120:
    #df.drop(x, inplace = True)

Yinelenen Kayıtları Bulma
Yinelenen satırlar, birden fazla kez kaydedilmiş satırlardır.


      Süre Tarih Nabız Maksimum Nabız Kalori
  0 60 '2020/12/01' 110 130 409.1
  1 60 '2020/12/02' 117 145 479.0
  2 60 '2020/12/03' 103 135 340.0
  3 45 '2020/12/04' 109 175 282.4
  4 45 '2020/12/05' 117 148 406.0
  5 60 '2020/12/06' 102 127 300.0
  6 60 '2020/12/07' 110 136 374.0
  7 450 '2020/12/08' 104 134 253.3
  8 30 '2020/12/09' 109 133 195.1
  9 60 '2020/12/10' 98 124 269.0
  10 60 '2020/12/11' 103 147 329.3
  11 60 '2020/12/12' 100 120 250.7
  12 60 '2020/12/12' 100 120 250.7
  13 60 '2020/12/13' 106 128 345.3
  14 60 '2020/12/14' 104 132 379.3
  15 60 '2020/12/15' 98 123 275.0
  16 60 '2020/12/16' 98 120 215.2
  17 60 '2020/12/17' 100 120 300.0
  18 45 '2020/12/18' 90 112 NaN
  19 60 '2020/12/19' 103 123 323.0
  20 45 '2020/12/20' 97 125 243.0
  21 60 '2020/12/21' 108 131 364.2
  22 45 NaN 100 119 282.0
  23 60 '2020/12/23' 130 101 300.0
  24 45 '2020/12/24' 105 132 246.0
  25 60 '2020/12/25' 102 126 334.5
  26 60 20201226 100 120 250.0
  27 60 '2020/12/27' 92 118 241.0
  28 60 '2020/12/28' 103 132 NaN
  29 60 '2020/12/29' 100 132 280.0
  30 60 '2020/12/30' 102 129 380.3
  31 60 '2020/12/31' 92 115 243.0

Test verilerimize baktığımızda, 11. ve 12. satırların birbirinin aynısı olduğunu varsayabiliriz.

Yinelenen kayıtları bulmak için bu yöntemi kullanabiliriz duplicated().

Bu duplicated()yöntem her satır için Boolean değerler döndürür:

In [ ]:
#TrueYinelenen her satır için bir değer döndürür , aksi takdirde False:

#print(df.duplicated())

# **Yinelenenleri Kaldırma**
Yinelenen kayıtları kaldırmak için bu drop_duplicates()yöntemi kullanın.

In [ ]:
#Yinelenen tüm öğeleri kaldır:

#df.drop_duplicates(inplace = True)

# **İlişkiler Kurmak**
Pandas modülünün harika bir özelliği de metottur corr().

Bu corr()yöntem, veri setinizdeki her sütun arasındaki ilişkiyi hesaplar.

Bu sayfadaki örnekler 'data.csv' adlı bir CSV dosyası kullanmaktadır.

In [ ]:
#Sütunlar arasındaki ilişkiyi gösterin:

#df.corr()

# **Sonuç Açıklaması**
Bu yöntemin sonucu, corr()iki sütun arasındaki ilişkinin ne kadar iyi olduğunu gösteren birçok sayı içeren bir tablodur.

Sayı -1 ile 1 arasında değişir.

1, bire bir ilişki (mükemmel korelasyon) olduğu anlamına gelir ve bu veri kümesi için, birinci sütundaki bir değer her yükseldiğinde, diğer sütundaki değer de yükselmiştir.

0,9 da iyi bir ilişki oranıdır ve bir değeri artırırsanız, diğer değer de muhtemelen artacaktır.

-0.9 değeri de 0.9 kadar iyi bir ilişki olurdu, ancak bir değeri artırırsanız, diğeri muhtemelen düşer.

0,2 iyi bir ilişki olmadığını gösterir; yani bir değerin yükselmesi diğerinin de yükseleceği anlamına gelmez.

İyi bir korelasyon nedir? Kullanım amacına bağlıdır, ancak iyi bir korelasyon olarak adlandırmak için en az 0.6(veya ) olması gerektiğini söylemek güvenli olur diye düşünüyorum .-0.6

Mükemmel Korelasyon:
"Süre" ve "Süre" sütunlarının aynı sayıyı aldığını görüyoruz 1.000000, bu da mantıklı; her sütunun kendisiyle mükemmel bir ilişkisi var.

İyi Korelasyon:
"Süre" ve "Kalori" 0.922721arasında çok iyi bir korelasyon var ve şunu tahmin edebiliriz: Ne kadar uzun süre egzersiz yaparsanız, o kadar çok kalori yakarsınız ve bunun tersi de geçerlidir: eğer çok kalori yaktıysanız, muhtemelen uzun bir egzersiz yapmışsınızdır.

Kötü Korelasyon:
"Süre" ve "Maksimum Nabız" 0.009403arasında çok zayıf bir korelasyon var; yani sadece egzersizin süresine bakarak maksimum nabzı tahmin edemeyiz ve bunun tersi de geçerli.

# **Çizim**
plot()Pandas, diyagramlar oluşturmak için bu yöntemi kullanır .

Diyagramı ekranda görselleştirmek için Matplotlib kütüphanesinin bir alt modülü olan Pyplot'u kullanabiliriz.

Matplotlib hakkında daha fazla bilgiyi Matplotlib Eğitim Kılavuzumuzda bulabilirsiniz .

In [ ]:
#Matplotlib'den pyplot'u içe aktarın ve DataFrame'imizi görselleştirin:

#import pandas as pd
#import matplotlib.pyplot as plt

#df = pd.read_csv('data.csv')

#df.plot()

#plt.show()

# **Dağılım Grafiği**
Aşağıdaki argümanı kullanarak dağılım grafiği istediğinizi belirtin kind:

kind = 'scatter'

Bir dağılım grafiğinin x ve y eksenine ihtiyacı vardır.

Aşağıdaki örnekte x ekseni için "Süre"yi, y ekseni için ise "Kalori"yi kullanacağız.

x ve y argümanlarını şu şekilde ekleyin:

x = 'Duration', y = 'Calories'

In [40]:
#import pandas as pd
#import matplotlib.pyplot as plt

#df = pd.read_csv('data.csv')

#df.plot(kind = 'scatter', x = 'Duration', y = 'Calories')

#plt.show()

Hatırlayın: Önceki örnekte, "Süre" ve "Kalori" arasındaki ilişkinin olduğunu öğrenmiş 0.922721ve daha uzun sürenin daha fazla kalori yakılması anlamına geldiği sonucuna varmıştık.

Dağılım grafiğine bakarak katılıyorum.

Şimdi, "Süre" ve "Maksimum Nabız" gibi sütunlar arasında kötü bir ilişki bulunan başka bir dağılım grafiği oluşturalım ve korelasyonu şu şekilde hesaplayalım 0.009403:

In [ ]:
#Sütunlar arasında hiçbir ilişkinin bulunmadığı bir dağılım grafiği:

#import pandas as pd
#import matplotlib.pyplot as plt

#df = pd.read_csv('data.csv')

#df.plot(kind = 'scatter', x = 'Duration', y = 'Maxpulse')

#plt.show()

# **Histogram**
kindHistogram istediğinizi belirtmek için şu argümanı kullanın :

kind = 'hist'

Bir histogram için yalnızca bir sütun yeterlidir.

Bir histogram bize her aralığın sıklığını gösterir; örneğin, 50 ile 60 dakika arasında süren kaç antrenman oldu?

Aşağıdaki örnekte histogram oluşturmak için "Süre" sütununu kullanacağız:

Örnek

In [ ]:
#df["Duration"].plot(kind = 'hist')

Not: Histogram bize 50 ile 60 dakika arasında süren 100'den fazla antrenman olduğunu gösteriyor.